src/modules/ekf2/EKF/zero_gyro_update.cpp

```diff
@@ -0,0 +1,82 @@
+/****************************************************************************
+ *
+ *   Copyright (c) 2023 PX4 Development Team. All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ *
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in
+ *    the documentation and/or other materials provided with the
+ *    distribution.
+ * 3. Neither the name PX4 nor the names of its contributors may be
+ *    used to endorse or promote products derived from this software
+ *    without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ * FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ * COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ * BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS
+ * OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED
+ * AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ * ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ *
+ ****************************************************************************/
+
+/**
+ * @file zero_gyro_update.cpp
+ * Control function for ekf zero gyro update
+ */
+
+#include "ekf.h"
+
+void Ekf::controlZeroGyroUpdate(const imuSample &imu_delayed)
+{
+	// Fuse zero gyro at a limited rate
+	const bool zero_gyro_update_data_ready = isTimedOut(_time_last_zero_gyro_fuse, (uint64_t)2e5);
```

Comment:
I'm just curious if anything particular thought goes into the specific rate limiting timing? 

Wouldn't it be fair to simply let this fuse every IMU sample while still? Overkill and expensive, but avoid aliasing and downsampling considerations?

Reply:
Yes, I don't feel that fusing this at a full rate is really reasonable: the fusion process is quite expensive. So far I didn't do the downsampling because I'm actually not that scared by aliasing in this specific case where it is only active when "at rest". Because we're "at rest" we know that there is almost no noise on the signal (nothing more than the noise of the sensor itself that is smaller than the bias and random enough to not have aliasing issues).

I can try to run it at full speed and see if there is a significant increase in CPU load.

Reply:
Fusing at full speed increases the CPU load by 2% on fmu-v5, this is not acceptable. 

Now the question is: Is aliasing a real concern in this specific case?

Reply:
2% per EKF2 instance or across multiple?

> Now the question is: Is aliasing a real concern in this specific case?

My guess would be *mostly* not, but I bet if any exceptions slip through it's going to be a fun surprise for the end user.

I'm trying to think of even a contrived case like a nearby cooling fan that's in bad shape, but somehow not enough to disrupt vehicle at reset detection?

Maybe we could maintain and share down averaged IMU with drag fusion?

Reply:
> Maybe we could maintain and share down averaged IMU with drag fusion?

Nevermind, I see you already did it.

Reply:
> 2% per EKF2 instance or across multiple?

Per instance.

> Nevermind, I see you already did it.

Yes, I figured out that it would take more time and risk to find out if it's worth doing it than just doing it.